### ISSUE TITLE
Not generating graph images via self-hosted Kroki installation via docker on Linux (bare-metal) or docker-on-WSL2

### ISSUE THREAD

**Issue by willfe** [asciidoctor/asciidoctor-vscode#628]
### TL;DR

The extension isn't rendering graphs from a self-hosted Kroki instance running via docker-compose (either from a native Linux host or a local WSL2 instance running Debian). My best guess is I've either misconfigured the extension somehow or it's ignoring (or having trouble with) settings to make it talk to a self-hosted Kroki instance.

### Details

Kroki is running in on my WSL2 instance using docker-compose and is working normally. Tests from Windows and WSL clients both work using identical URLs. PNG and SVG formats are both working. In VSCode, with `kroki-server-url` set to `http://localhost:8000` and `asciidoc.extensions.enableKroki` set to `true`, all that appears for any graph (of any type) in the preview window is the "broken image" icon and the word "diagram" next to it. I've tried this with the current production release of the extension and the preview release (changing settings accordingly) with no change. I tested Kroki docker images ranging from `latest` back to `1.7.0` (three releases), also yielding no changes.

Annoyingly, the production/hosted Kroki at kroki.io seems to be having trouble right now (as it's not rendering Mermaid graphs there either) so even when I remove `kroki-server-url` from my configuration (to allow fallback to kroki.io), it just produces "Error 500: Ok" (lol) instead of the broken image icon.

Sample working URLs (localhost):

- [graphviz](http://localhost:8000/graphviz/svg/eNpLyUwvSizIUHBXqPZIzcnJ17ULzy_KSanlAgB1EAjQ)
- [mermaid](http://localhost:8000/mermaid/svg/eNqFUstugzAQvOcrfKroAfUHKi59q42UPqQerKhawAFEsNNd0_Tz67UpMfSRCyN2Z3Z2FiqEXS0enhZCvPeqVzK5-lDanud4lj1y4XQt0jQTYC0UdedaRMe4iqtHWRZBExS2MZq51OeV32VvsFVI8jWgOBErNIUiMkjrmNjoSpGVdx64E-3o3sYtZkZKl1Ou36aFTQujalKK9ZOGR5ncMwyhCsfpO_U2mv_D8XN_p2wM7gHLOO2gkhcB3VmuA4mTT22lPzjFDe8lX6Ik3B1s5MqQTYcr--8UjQ7nmhqI-WRxGCb-0nCuDhr9Qxs1vmdEJQaZLN3Tb_ZsDULl_qH4NrUxrUv9aRVq2IpbfmVC1dgt5NmNBy6UDRUGy-wy4CFfbBxUs-KgXHwBQQYTWg==)

Side note: I also tried setting up Kroki on a different host on my network and verified it was working (same URLs as above, just with its hostname there instead of "localhost"), but pointing my VSCode configuration to that base URL didn't change things either. 

### VSCode
```
Version: 1.70.2
Commit: e4503b30fc78200f846c62cf8091b76ff5547662
Date: 2022-08-16T05:35:13.448Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Windows_NT x64 10.0.19044
```

`settings.json` section for AsciiDoc extension (3.0.1 pre-release version):
```json
"asciidoc.extensions.enableKroki": true,
"asciidoc.preview.asciidoctorAttributes": {
  "kroki-server-url": "http://localhost:8000"
}
```

`settings.json` section for AsciiDoc extension (latest production release):
```json
"asciidoc.use_kroki": true,
"asciidoc.preview.attributes": {
  "kroki-server-url": "http://odin:8000"
},
```

### WSL2
```powershell
$ wsl --status
Default Distribution: Debian
Default Version: 2

Windows Subsystem for Linux was last updated on 3/27/2022
WSL automatic updates are on.

Kernel version: 5.10.102.1
```

### Debian instance
```console
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
$ uname -a
Linux POSEIDON 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 GNU/Linux
```

### "Bare metal" Linux host
Synology DS2419+
DSM 7.1-42661 Update 4
```console
$ uname -a
Linux odin 4.4.180+ #42661 SMP Mon Jun 27 15:09:45 CST 2022 x86_64 GNU/Linux synology_denverton_2419+
```
Note: I didn't test with this host for very long since it didn't change anything -- Kroki worked fine on this host as well, but the VSCode extension still just showed the same symptom. I thought it was worth trying it out on this "esoteric" Linux distribution, figuring if Kroki can run on this thing via docker, it can run on pretty much anything that docker can. Didn't help with this issue though. 😢 

### Extension
AsciiDoc 3.0.1 Pre-Release
Last updated: 7/15/2022, 11:10:02
`Extension is enabled on 'WSL: Debian'`

### Docker
```console
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  compose: Docker Compose (Docker Inc., v2.2.3)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 5
  Running: 5
  Paused: 0
  Stopped: 0
 Images: 48
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 0197261a30bf81f1ee8e6a4dd2dea0ef95d67ccb
 runc version: v1.1.3-0-g6724737
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.102.1-microsoft-standard-WSL2
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 15.62GiB
 Name: POSEIDON
 ID: HHPC:RYTP:YM6J:EWTW:OZFE:D3U6:ZLOY:VQGD:FAEX:UNDW:WV5K:36FH
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: API is accessible on http://0.0.0.0:2375 without encryption.
         Access to the remote API is equivalent to root access on the host. Refer
         to the 'Docker daemon attack surface' section in the documentation for
         more information: https://docs.docker.com/go/attack-surface/
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

docker_compose.yml:
```yaml
version: "3"
services:
  kroki:
    image: yuzutech/kroki
    depends_on:
      - blockdiag
      - mermaid
      - bpmn
      - excalidraw
    environment:
      - KROKI_BLOCKDIAG_HOST=blockdiag
      - KROKI_MERMAID_HOST=mermaid
      - KROKI_BPMN_HOST=bpmn
      - KROKI_EXCALIDRAW_HOST=excalidraw
      - KROKI_SAFE_MODE=unsafe
      - KROKI_MAX_URI_LENGTH=30000
    ports:
      - "8000:8000"
  blockdiag:
    image: yuzutech/kroki-blockdiag
    ports:
      - "8001:8001"
  mermaid:
    image: yuzutech/kroki-mermaid
    ports:
      - "8002:8002"
  bpmn:
    image: yuzutech/kroki-bpmn
    ports:
      - "8003:8003"
  excalidraw:
    image: yuzutech/kroki-excalidraw
    ports:
      - "8004:8004"
```

## To Reproduce

Prerequisites:
- Working WSL2 installation running a Debian or Ubuntu instance.
- Working Docker installation running in the instance, including docker-compose.
- Visual Studio Code installation with WSL integration enabled and running normally (docker support/features are not required).

Steps:

1. Launch Kroki Inside the WSL2 instance:
    1. `mkdir -p ~/tmp/kroki`
    2. `cd ~/tmp/kroki`
    3. Create `docker-compose.yml` with the content shown above
    4. `docker-compose up -d`
2. Once docker-compose finishes bringing up kroki, verify functionality:
    1. Click the sample URLs provided above to verify functionality (this verifies it's working from your Windows environment)
    2. To verify functionality Inside the WSL2 instance:
      ```console
      $ wget -O mermaid.svg [sample mermaid URL from above]
      $ file mermaid.svg
      mermaid.svg: SVG Scalable Vector Graphics image
      ```
3. Install and activate the AsciiDoc extension in VSCode.
4. Add the following to `settings.json` (user prefs, just to simplify things for testing)
    ```json
    "asciidoc.extensions.enableKroki": true,
      "asciidoc.preview.asciidoctorAttributes": {
      "kroki-server-url": "http://localhost:8000"
    }
    ```
5. Create a document named `test.asciidoc` or `test.adoc`:
    ```asciidoc
    Test document.

    [mermaid]
    ....
    graph TD
        T1[Test 1]
        T1 -->|true| T1.1[Test 1.1]
    ....
    ```
6. Open the preview pane for the document and scowl menacingly at that broken image icon.

I'm open to any suggestions to resolve this and I'm at your disposal for testing and feedback.

Thanks!

**Comment by Mogztter** [comment]
Wow, thanks for the very detailed issue!
Did you change the `Asciidoc: Change Preview Security Settings` to allow the local content to load?

Regarding your comment, Kroki.io displays an error image when the accept type is contains html. Not sure exactly why the error message is `Ok` though 😄

We had a few issues with Mermaid lately. I'm currently trying to pinpoint what is happening but the consequence is that the communication with the Chromium is lost and as a result we cannot render Mermaid diagrams anymore.

**Comment by willfe** [comment]
> Did you change the `Asciidoc: Change Preview Security Settings` to allow the local content to load?

I'm not sure where to make that change. I searched for various parts of that phrase in VSCode's settings panel and didn't see anything that comes close to matching. The only security-related change I've made is setting `KROKI_SAFE_MODE=unsafe` inside the docker configuration for the Kroki container.

**Comment by Mogztter** [comment]
It's a VS Code command:

<img width="1090" alt="Capture d’écran 2022-08-28 à 15 29 16" src="https://user-images.githubusercontent.com/333276/187076378-0f30c9d7-17c0-438b-b2c4-d74b416abbf0.png">

As an enhancement, we could automatically allow content from the configured `kroki-server-url`.

**Comment by willfe** [comment]
lol that did the trick! Thank you for the pointer. I'd have never spotted that otherwise.

**Comment by Mogztter** [comment]
> As an enhancement, we could automatically allow content from the configured kroki-server-url.

If you think that's a good idea or if you have a better idea on how to improve the situation feel free to open a new issue. Maybe we just need to improve the documentation?

Thanks!

**Comment by willfe** [comment]
From a security perspective, it's probably best to leave it as-is so that the user is still making a conscious decision (and taking explicit action) to enable that trust setting. IMHO it's worth noting in the docs as a required step for using a hosted Kroki instance, but not auto-enabling trust when `kroki-server-url` is present.

**Comment by Mogztter** [comment]
Thanks for your feedback.

The only issue is that, currently, it's either enable unsecure (http) on all hosts or not. In other words, we cannot enable a particular host using the "Manage Preview Security Settings", for instance: `http://odin:8000`.

Since we can define a CSP rule on a specific protocol, host and port (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#sources) I think it's fine to automatically add a rule when `enableKroki` is true.

Please note that using a remote server (https://kroki.io) is allowed by default. So it seems fair to also allow `kroki-server-url` by default, no?

**Comment by willfe** [comment]
Oh, that's a good point. You're right, enabling it for user-specified URLs makes sense to me now since it's already allowed for the official Kroki service anyway.

**Comment by haydencbarnes** [comment]
@Mogztter Was this "allow `kroki-server-url` by default" or documentation ever implemented? I just followed basically the same setup as 'willfe' here, setting up a self-hosted Kroki instance utilizing all the documentation available and multiple git issues, but the only issue was the "Asciidoc: Change Preview Security Settings" option needing to be set correctly to allow the local content to load.

**Comment by Mogztter** [comment]
> Was this "allow kroki-server-url by default" or documentation ever implemented?

No, feel free to open a new issue to request this change/enhancement 😄
